File: collect_app/src/androidTest/java/org/odk/collect/android/support/ScreenshotOnFailureTestRule.java
package org.odk.collect.android.support;

import org.junit.rules.TestWatcher;
import org.junit.runner.Description;

import tools.fastlane.screengrab.Screengrab;
import tools.fastlane.screengrab.UiAutomatorScreenshotStrategy;

public class ScreenshotOnFailureTestRule extends TestWatcher {

    @Override
    protected void failed(Throwable e, Description description) {
        super.failed(e, description);
        takeScreenshot(description);
    }

    private void takeScreenshot(Description description) {
        Screengrab.setDefaultScreenshotStrategy(new UiAutomatorScreenshotStrategy());

        String filename = description.getTestClass().getSimpleName() + "-" + description.getMethodName();
        Screengrab.screenshot(filename);
    }
}

File: collect_app/src/androidTest/java/org/odk/collect/android/support/rules/ScreenshotOnFailureTestRule.kt
package org.odk.collect.android.support.rules

import org.junit.rules.TestWatcher
import org.junit.runner.Description
import tools.fastlane.screengrab.Screengrab
import tools.fastlane.screengrab.UiAutomatorScreenshotStrategy

class ScreenshotOnFailureTestRule : TestWatcher() {

    override fun failed(e: Throwable, description: Description) {
        super.failed(e, description)
        takeScreenshot(description)
    }

    private fun takeScreenshot(description: Description) {
        Screengrab.setDefaultScreenshotStrategy(UiAutomatorScreenshotStrategy())

        val filename = description.testClass.simpleName + "-" + description.methodName
        Screengrab.screenshot(filename)
    }
}
